### PR TITLE
fix(api): Add pagination for channel.list in Slack integration

### LINKED ISSUE #10056: New global Slack integration only pulls back 1000 channels when finding the channel ID
## Important Details

How are you running Sentry?

* [X] On-Premise docker [Version 9.0.0]
* [ ] Saas (sentry.io)
* [ ] Other [briefly describe your environment]

## Description

I was attempting to add in Slack integration using the new global integration. I tried to add one of our channels, but it kept saying it "does not exist or has not been granted access"

After poking around for a while, I realized that we were only getting back 1000 channels, and that there is token-based pagination happening per https://api.slack.com/methods/channels.list

## Steps to Reproduce

1. Integrate with a Slack instance with > 1000 channels
2. Attempt to add an alert to a channel that is after the first 1000 channels added
3. See it fail

### What you expected to happen

I would expect it to paginate until all channels are returned

### Possible Solution

I'm working right now on implementing basic pagination to get this working for us in the short term. I will update this task if I get something put together.

### PR DESCRIPTION
**Pull request by cjmaio**
In the global Slack integration, `channel.list` is used to grab the channel ID. However, in Slack instances with > 1000 channels, this is insufficient. This change will fire off subsequent requests if needed to fetch the entire list of channels.

Fixes GH-10056

### CODE REVIEW

File: src/sentry/integrations/slack/notify_action.py [hunk 1]
```diff
@@ -182,13 +182,21 @@ def get_channel_id(self, integration_id, name):
...
+            }))
+            resp = resp.json()
+            if not resp.get('ok'):
+                self.logger.info('rule.slack.channel_list_failed', extra={'error': resp.get('error')})
+                return None
+            channels.extend(resp['channels'])
+            cursor = resp.get('response_metadata', {}).get('next_cursor', None)
 
-        channel_id = {c['name']: c['id'] for c in resp['channels']}.get(name)
+        channel_id = {c['name']: c['id'] for c in channels}.get(name)
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/slack/notify_action.py]: Hi @cjmaio! Sorry for the delay on this. Yes, we should paginate.

Can we change it to check if the reuqested channel name exists after each page, that way we don't needlessly continue paging when we could have found the ID in the first page (for example).

**Review by evanpurkhiser — CHANGES_REQUESTED** [review]

File: src/sentry/integrations/slack/notify_action.py [hunk 2]
```diff
@@ -182,16 +182,22 @@ def get_channel_id(self, integration_id, name):
...
         })
 
-        resp = session.get('https://slack.com/api/channels.list', params=channels_payload)
-        resp = resp.json()
-        if not resp.get('ok'):
-            self.logger.info('rule.slack.channel_list_failed', extra={'error': resp.get('error')})
-            return None
+        # Slack limits the response of `channels.list` to 1000 channels, paginate if needed
+        cursor = None
+        while cursor or len(channels) == 0:
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/slack/notify_action.py]: Where is `channels` coming from here?

**Comment by cjmaio** [comment]
Closing this, something weird happened... will reopen.

File: src/sentry/integrations/slack/notify_action.py (hunk 2 above)
Comment (cjmaio) [reply src/sentry/integrations/slack/notify_action.py]: Sorry, that was from a bad merge. I resolved this concern and put up a new PR https://github.com/getsentry/sentry/pull/11064

**Pull request closed without merging**